go/worker/storage/committee/checkpoint_sync.go

```diff
@@ -0,0 +1,390 @@
+	var outChan chan *checkpoint.ChunkMetadata
+
+	for {
+		if chunks.length == 0 {
+			next = nil
+			outChan = nil
+		} else {
+			next = heap.Pop(chunks).(*checkpoint.ChunkMetadata)
+			outChan = chunkDispatchCh
+		}
+
+		select {
+		case <-n.ctx.Done():
+			return checkpointStatusBail, n.ctx.Err()
+
+		case returned := <-chunkReturnCh:
+			if returned == nil {
+				// Restoration completed, no more chunks.
+				return checkpointStatusDone, nil
+			}
+			heap.Push(chunks, returned)
+
+		case status := <-errorCh:
+			return status, nil
+
+		// If there's no chunk to send, outChan will be nil here, blocking forever. We still need to wait
+		// for other events even if there's no chunk to dispatch, since they may simply all be in processing.
+		case outChan <- next:
+			next = nil
+
+		case <-doneCh:
+			// No usable committee connections left, move on to the next checkpoint.
+			return checkpointStatusNext, storageClient.ErrStorageNotAvailable
+		}
+
+		if next != nil {
+			heap.Push(chunks, next)
+		}
+	}
+}
+
+func (n *Node) getCheckpointList(committeeClient committee.Client) ([]*checkpoint.Metadata, error) {
+	// Get checkpoint list from all current committee members.
+	listCh := make(chan []*checkpoint.Metadata)
+	req := &checkpoint.GetCheckpointsRequest{
+		Namespace: n.commonNode.Runtime.ID(),
+	}
+	getter := func(ctx context.Context, conn *committee.ClientConnWithMeta) error {
+		api := storageApi.NewStorageClient(conn.ClientConn)
+		meta, err := api.GetCheckpoints(ctx, req)
+		if err != nil {
+			return err
+		}
+		n.logger.Info("got checkpoint list from a node", "length", len(meta))
+		listCh <- meta
+		return nil
+	}
+
+	cancel, doneCh, err := n.goWithCommittee(committeeClient, getter)
+	if cancel != nil {
+		defer cancel()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var list []*checkpoint.Metadata
+resultLoop:
+	for {
+		select {
+		case <-doneCh:
+			break resultLoop
+		case <-n.ctx.Done():
+			return nil, n.ctx.Err()
+		case meta := <-listCh:
+			list = append(list, meta...)
+		}
+	}
+
+	// Prepare the list: sort and deduplicate.
+	sort.Slice(list, func(i, j int) bool {
+		// Descending!
+		if list[j].Root.Version == list[i].Root.Version {
+			if list[j].Root.Namespace.Equal(&list[i].Root.Namespace) {
+				return bytes.Compare(list[j].Root.Hash[:], list[i].Root.Hash[:]) < 0
+			}
+			return bytes.Compare(list[j].Root.Namespace[:], list[i].Root.Namespace[:]) < 0
+		}
+		return list[j].Root.Version < list[i].Root.Version
+	})
+	retList := make([]*checkpoint.Metadata, len(list))
+	var prevCheckpoint *checkpoint.Metadata
+	cursor := 0
+	for i := 0; i < len(list); i++ {
+		if prevCheckpoint == nil || !list[i].Root.Equal(&prevCheckpoint.Root) {
+			retList[cursor] = list[i]
+			cursor++
+		}
+	}
+
+	return retList[:cursor], nil
+}
+
+func (n *Node) checkCheckpointUsable(cp *checkpoint.Metadata) bool {
+	if n.syncedState.LastBlock.IORoot.Namespace.Equal(&cp.Root.Namespace) && n.syncedState.LastBlock.IORoot.Version >= cp.Root.Version {
```

Comment:
Doesn't accessing `n.syncedState` require holding a lock (if caller should lock, rename the method to `...Locked`)? Also it would be more clear if namespace comparisons were against ` n.commonNode.Runtime.ID()`?